Rectangle struct

vancouver29 opened the pull request:
devide a rectangle to 4 quadrants

vancouver29 committed:
Rectangle struct

devide a rectangle to 4 quadrants

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -119,8 +119,60 @@ impl Vector {
+// Rectangle /////////////////////////////////////////////////////////////////
+pub struct Size{
+    pub width: f32,
+    pub heigh: f32,
```
Typo: height

Also, there should be a space between `Size` and `{`

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -119,8 +119,60 @@ impl Vector {
+    pub heigh: f32,
+}
+
+pub struct Rectangle {
```
I would suggest to rename `Rectangle` to `Rect`. It's always better to use shorter names when possible so that the code remains clean.

vancouver29 committed:
Bracket is fixed and clone methods is added

Bracket is fixed and clone methods is added

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -119,8 +119,60 @@ impl Vector {
+    }
+
+    pub fn quadrants(&self) -> Vec<Rectangle> {
+        let mut v:Vec<Rectangle> = Vec::new();
```
I would suggest not using a `Vec` here: we are always going to return four `Rect` objects, so we can just return then in a tuple. This gives us the benefit of restructuring, like this:

`let (a, b, c, d) = some_rect.quadrants();`

vancouver29 replied:
so the return type should be (Rect,Rect,Rect,Rect) ?

johnxnguyen replied:
Yes, but don't forget the spaces! Like this `(Rect, Rect, Rect, Rect)`

vancouver29 committed:
Rectangle to Rect

Abbreviation

vancouver29 committed:
space between Size and { is fixed

space between Size and {
space between Rect and {

vancouver29 committed:
typo heigh -> height is fixed

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +121,73 @@ impl Vector {
+    pub fn half_size(&mut self) {
+        self.size.width = self.size.width/2.0;
+        self.size.heigh = self.size.heigh/2.0;
+    }
```
Here's an idea to make the code more concise. We could replace this method `half_size()` with a private instance method that returns a new `Rect` with the size already smaller. 

For example: `fn quarter_sized(&self) -> Rect`

The origin would still be the same, but the size is already smaller. Then in the `quadrants` method you wouldn't need to use `clone()` anymore (you could even delete the implementation of `clone()`), you could just write:

`let mut top_left = self.quarter_sized();`

vancouver29 replied:
its a good idea, i will think about it now

vancouver29 replied:
i just thought about this idea, the size of the rectangle can be smaller, but if we use this method, it is just true for  a quarter of the rectangle with the coordinator , because the other 3 Rectangles have different coordinators

vancouver29 replied:
ah we still can change the coordinator extra ^^

vancouver29 committed:
Test function rect_quadrant()

there are no errors , but im wodering why there are no test analysing in terminal as usual

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +122,89 @@ impl Vector {
+impl Rect {
+
+    pub fn half_size(&mut self) {
+        self.size.width = self.size.width/2.0;
```
We don't need this method any more, it's not being used so you can delete it.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +122,89 @@ impl Vector {
+                x: self.origin.x,
+                y: self.origin.y,
+            },
+            size: Size{
```
You should be consistent with your whitespace, usually there is a space between a word and a {

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +122,89 @@ impl Vector {
+        let mut rect4 = self.quarter_sized();
+        rect4.origin.y = rect4.origin.y + rect4.size.height;
+
+        let rects: (Rect, Rect, Rect, Rect) = (rect1, rect2, rect3, rect4);
```
You don't need to specify the type here, the compiler can infer it. Also for clarity, I would suggest to give better names than `rect1` and `rect2`, something like `upper_left` and `upper_right`.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +122,89 @@ impl Vector {
+//    fn eq(&self, other: &Rect) -> bool {
+//        (self.origin == other.origin) && (self.size == other.size)
+//    }
+//}
```
You should delete this

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -121,6 +122,89 @@ impl Vector {
+
+
+impl Clone for Rect {
+    fn clone(&self) -> Self {
```
This can be deleted in you derive the `Clone` trait for `Rect`.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -230,4 +314,26 @@ mod tests {
+    fn rect_quadrant() {
+        // given
+        let mut rect = Rect { origin: Point { x: 0.0, y: 0.0 }, size: Size { width: 6.0, height: 8.0 } };
+
```
It's good practice to write this over several lines. When you write it inline like thus it is hard to read.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -230,4 +314,26 @@ mod tests {
+        // when
+        let result = rect.quadrants();
+
+        let (rect1, rect2, rect3, rect4) = result;
```
Like above, give these better names. Also you don't need to store it in result, you can destructure the return value of `quadrants()` directly.

vancouver29 committed:
Rect_names are fixed

ex: upper_left, upper_right, lower_right, lower_left

vancouver29 committed:
destructure rect

destructure rect.quadrants()

johnxnguyen reviewed (approved):
The styling is still a bit off, but I'll merge this now and make a new PR showing you how to settle your code.
